---
output: html_document
---

###Sea Cucumber Differential Gene Expression Data of Intestine Regenrative Tissue
**Joshua Medina** | 7/16/2019

```{r, echo=FALSE, message=FALSE, cache=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

###Overview

The sea cucumber _Holothuria glaberrima_ has great regenerative capability after internal organs evisceration. In this case, we are analyzing data from mesentery regenerative tissue of 1 and 3 days post evisceration (1- and 3DPE) in order to further understand the genes that are involved in the early stages of regeneration.

***

####Process

Set the directory to work in and load required libraries.

```{r}

setwd("~/git/DEdq2019/")

suppressMessages(library(DESeq2))
suppressMessages(library(gplots))
suppressMessages(library(RColorBrewer))
suppressMessages(library(EnhancedVolcano))
suppressMessages(library(tidyverse))

```


**Loading the data.**

* The variable *names* will contain the genenames or in this case the ID given by the assembler for all the reads in the count data to be used.
* The variable *ddata* contains a csv file indicating the files to be used and to which group they correspond (i.e experimental or control). In our case this is going to consist of three different levels normal, 1 day post-evisceration (1DPE), and 3 days post-evisceration (3DPE).

```{r}

names <- read.table("data/genenames.txt", header=FALSE, row.names = 1)
ddata <- read.csv("data/ddata.csv", header = TRUE)

```

Generating the data set with respect to the counts.

```{r}
ddsSTAR <- DESeqDataSetFromHTSeqCount(sampleTable = ddata, directory = "data/countsmay2019", design = ~ condition)

colData(ddsSTAR)$condition <- factor(colData(ddsSTAR)$condition, levels=levels(ddata$condition))

#set a factor variable with six levels (conditions of the study). 
#Totally needed another option is to use relevel.

```

Estimating the size factors using _"median ratio method"_, dispersions (gene-wise, mean-dispersion relationship, final dispersion estimates). Generating the counts table with _counts_ function.

```{r}

dds <- estimateSizeFactors(ddsSTAR)
dds <- estimateDispersions(dds)
dds <- nbinomWaldTest(dds)
counts_table <- counts(dds, normalized=TRUE)

```

Restults of contrasts between conditions (3DPE vs Normal; 1DPE vs Normal). __Note: Normal tissue is referred to non-regenerated mesentery tissue of the sea cucumber.__

```{r}

res_3DPE <- results(dds, contrast = c("condition", "3DPE", "Control"))
res_3DPE <- res_3DPE[order(res_3DPE$padj),]

write.csv(as.data.frame(res_3DPE), file="data/dqData2019_3DPE.csv")
```

####Visualization of the results.

***

Plotting dispersion estimates.

```{r, fig.align = 'center'}
DESeq2::plotDispEsts(dds,main="Dispersion Plot")
```

Raw Data vs normalized counts.

```{r, fig.align = 'center'}

conds <- as.vector(ddata$condition)
condcols <- brewer.pal(n = length(unique(conds)), name="Dark2") 
par(mfrow=c(1,2))
barplot(colSums(counts(dds, normalized=F)), col = condcols[as.factor(conds)], las = 2, cex.names=0.4, main="Raw Counts")
barplot(colSums(counts(dds, normalized=T)), col=condcols[as.factor(conds)], las=2, cex.names = 0.4, main = "Normalized Counts")

#is normalizing really a good thing for this data?

```


PCA Plot

```{r, fig.align = 'center'}

vsd <- varianceStabilizingTransformation(dds)
DESeq2::plotPCA(vsd)+theme_bw()

```

Volcano Plot

* The purpose of the code in this section is to assign colors and identification specific to the fold change of each transcript. The base for this will be a fold change of 1, transcripts with fold change of 1.0 or higher will have a royalblue color, those with fold change of -1.0 or lower will have red and those in between will be black.

```{r}

keyvals <- rep("black", nrow(res_3DPE))

names(keyvals) <- rep('Mid', nrow(res_3DPE))

keyvals[which(res_3DPE$log2FoldChange > 1.0)] <- 'royalblue'
names(keyvals)[which(res_3DPE$log2FoldChange > 1.0)] <- 'high'

keyvals[which(res_3DPE$log2FoldChange < -1.0)] <- 'red'
names(keyvals)[which(res_3DPE$log2FoldChange < -1)] <- 'low'

```

Removing duplicates from keyvals.

```{r}

unique(names(keyvals))

unique(keyvals)

head(keyvals)
tail(keyvals)

```

The actual plot... 

```{r, fig.align = 'center'}

EnhancedVolcano(res_3DPE,
                lab = rownames(res_3DPE),
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c('TRINITY_DN57277_c0_g3_i1'),
                xlim = c(-14,14),
                xlab = bquote(~Log[2]~ 'fold change'),
                pCutoff = 10e-2,
                FCcutoff = 1.0,
                transcriptPointSize = 1.0,
                transcriptLabSize = 3.0,
                transcriptLabCol = 'black',
                transcriptLabFace = 'bold',
                boxedlabels = TRUE,
                colCustom = keyvals,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 14,
                legendIconSize = 4.0,
                drawConnectors = TRUE,
                widthConnectors = 0.75,
                colConnectors = 'black',
                gridlines.major = FALSE,
                gridlines.minor = FALSE)

```

Heatmap of 100 samples...

```{r}

hmcol <- colorRampPalette(brewer.pal(9, "YlOrBr"))(100)
hits_3DPEvsNORM <- rownames(res_3DPE[(res_3DPE$padj <= 0.001) & (!is.na(res_3DPE$padj)) & (abs(res_3DPE$log2FoldChange) >= 2),])
hits_3DPEvsNORM2 <- hits_3DPEvsNORM[1:100]
heatmap.2(assay(vsd[hits_3DPEvsNORM2,]), trace="none", col=hmcol, cexRow = 0.2, cexCol = 0.6, las=2, Colv=TRUE, dendogram="row", key=TRUE, main="3DPE vs Normal")


```

